### PR TITLE
ARROW-16223: [C++] Fix decimal reduce scale rounding

### PR DESCRIPTION
**Pull request by cyb70289**

### CODE REVIEW

File: cpp/src/arrow/util/decimal_test.cc [hunk 1]
```diff
@@ -1521,6 +1525,10 @@ TEST(Decimal128Test, ReduceScaleAndRound) {
   result = Decimal128("-123750").ReduceScaleBy(2, true);
   ASSERT_OK(result.ToInteger(&out));
   ASSERT_EQ(-1238, out);
+
+  result = Decimal128("-5").ReduceScaleBy(1, true);
+  ASSERT_OK(result.ToInteger(&out));
+  ASSERT_EQ(-1, out);
```
Comment (pitrou) [inline cpp/src/arrow/util/decimal_test.cc:1531]: Can we also add tests for `Decimal128(0)`?

Comment (cyb70289) [reply cpp/src/arrow/util/decimal_test.cc:1535]: Done

**Comment by pitrou** [comment]
@lidavidm Are the Flight compile failures on MinGW known here?

**Review by pitrou — APPROVED** [review]
+1, the CI issues look unrelated

**Comment by lidavidm** [comment]
> @lidavidm Are the Flight compile failures on MinGW known here?

Hmm, no, that's new. I filed ARROW-16235. I'll look at it and ARROW-16215 today.

**Pull request closed without merging**